django_ai_assistant/helpers/assistants.py

```diff
@@ -437,6 +449,115 @@ def get_history_aware_retriever(self) -> Runnable[dict, RetrieverOutput]:
             prompt | llm | StrOutputParser() | retriever,
         )
 
+    @with_cast_id
+    def as_graph(self, thread_id: Any | None = None) -> Runnable[dict, dict]:
+        """Create the Langchain graph for the assistant.\n
+        This graph is an agent that supports chat history, tool calling, and RAG (if `has_rag=True`).\n
+        `as_graph` uses many other methods to create the graph for the assistant.
+        Prefer to override the other methods to customize the graph for the assistant.
+        Only override this method if you need to customize the graph at a lower level.
+
+        Args:
+            thread_id (Any | None): The thread ID for the chat message history.
+                If `None`, an in-memory chat message history is used.
+
+        Returns:
+            the compiled graph
+        """
+        llm = self.get_llm()
+        tools = self.get_tools()
+        llm_with_tools = llm.bind_tools(tools) if tools else llm
+        message_history = self.get_message_history(thread_id)
+
+        def custom_add_messages(left: list[BaseMessage], right: list[BaseMessage]):
+            result = add_messages(left, right)
+
+            if thread_id:
+                messages_to_store = [
+                    m
+                    for m in result
+                    if isinstance(m, HumanMessage | ChatMessage)
+                    or (isinstance(m, AIMessage) and not m.tool_calls)
+                ]
+                message_history.add_messages(messages_to_store)
+
+            return result
+
+        class AgentState(TypedDict):
+            messages: Annotated[list[AnyMessage], custom_add_messages]
+            input: str  # noqa: A003
+            context: str
+            output: str
+
+        def setup(state: AgentState):
+            return {"messages": [SystemMessage(content=self.get_instructions())]}
+
+        def retriever(state: AgentState):
+            if not self.has_rag:
+                return
+
+            retriever = self.get_history_aware_retriever()
```

Comment:
Seems the `get_history_aware_retriever` is still using `RunnableBranch`, to summarize the previous conversation when has_rag = True and the user is making a second question. That's fine, but we may want to change this to use langgraph in the next PR. LMK if you need clarification on this.

Reply:
Yes, this was intentional to keep things simpler for now, definitely something we should migrate as well later

Reply:
Created an issue for it https://github.com/vintasoftware/django-ai-assistant/issues/166

Reply:
Thanks!